doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,979 @@
+
+.. code:: C++
+
+    class MyPass : public BasicBlockPass
+
+
+Do or do not there is no... test
+********************************
+
+I am sure that your are eager to compile and run this empty pass. Thanks to the files provided in the Quarkslab git it's actually quite easy.
+First you need to tell cmake that your pass should be compiled by adding it in the file ``$PASSDIR/llvm-passes/CMakeList.txt``.
+It should now look like this:
+
+.. code:: cmake
+
+    set(EPONA_LLVM_MODULES
+        ObfuscateZero
+        MyPass
+    )
+
+Now we are going to build the pass:
+
+.. code:: bash
+
+    >$ cd $PASSDIR
+    >$ mkdir build
+    >$ cd build
+    >$ cmake -DLLVM_ROOT=path/to/your/llvm/build ..
+    >$ make
+
+And now let's run our pass with clang. But we need a test file, write the following code somewhere:
+
+.. code:: c
+
+    #include <stdio.h>
+
+    int foo(){return 1;}
+
+    int main() {
+        printf("Hello world\n");
+
+        return 0;
+    }
+
+
+
+And compile it with our awesome pass using:
+
+.. code:: bash
+
+    >$ clang++ -Xclang -load -Xclang $PASSDIR/build/llvm-passes/LLVMMyPass.so path/to/test/file -o /tmp/awesome.out
+
+Or if you already have a file with the LLVM bytecode:
+
+.. code:: bash
+
+  >$ opt -load $PASSDIR/build/llvm-passes/LLVMMyPass.so -MyPass in.ll -o out.bc
+
+
+Since there are two blocks in our code (foo and main), we see the message "I m running on a block..." twice!
+
+Congratulations you have compiled your first program with an LLVM pass! (You can test the executable, it should work... should)
+
+Playtime is over
+****************
+
+The method we have to implement is ``runOnBasicBlock`` which takes as parameter a reference to the current block. We will proceed step by step.
+
+Finding null literals
++++++++++++++++++++++
+
+To find the null literals we need to iterate over every instruction of the block and check if one of the operands is null.
+
+.. code:: C++
+
+  //Add the following to your headers
+  #include "llvm/IR/Constants.h"
+  #include "llvm/IR/Instructions.h"
+
+  //Add the following to MyPass
+  virtual bool runOnBasicBlock(BasicBlock &BB) override {
+    // Not iterating from the beginning to avoid obfuscation of Phi instructions
+    // parameters
+    for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
+                                       end = BB.end();
+         I != end; ++I) {
+      Instruction &Inst = *I;
+      // We are not using an iterator because we will need i later.
+      for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
+        if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
+          errs() << "I've found one sir!\n";
+        }
+      }
+    }
+    return false;
+   }
+
+  Constant *isValidCandidateOperand(Value *V) {
+    if (Constant *C = dyn_cast<Constant>(V)) {
+      // Checking constant eligibility
+      if (isa<PointerType>(C->getType())) {
+        // dbgs() << "Ignoring NULL pointers\n";
```

Comment:
Oh I got it, these are "sanity checks".
I'd rather write:

```
if (!C->isNullValue() || 
    C->getType()->isFloatingPointTy() ||
    C->getType()->isPointerTy())
    return nullptr;
// with the other change below, all the check are done earlier and here 
// you are at the minimum indentation level. Easier control flow.

  ....

```

Note the use of ```C->getType()->isPointerTy()``` for consistency with the floating point check.